quickwit-storage/src/cache/disk/store.rs

```diff
@@ -0,0 +1,910 @@
+        // Although unlikely, we don't want to overwrite data we shouldn't be overwriting.
+        // this just prevents us from affecting data we dont want to touch.
+        let sub_slice = &bytes[0..range.len()];
+
+        let info = match self.write_range_inner(key, sub_slice, range.start as u64) {
+            Ok(info) => info,
+            Err(primary_error) => {
+                // We don't want to leave a random file laying around.
+                return if let Err(secondary_error) = self.remove_file(key) {
+                    Err(propagate_errors(primary_error, secondary_error))
+                } else {
+                    Err(primary_error)
+                };
+            }
+        };
+
+        let entry = self.register_file_write(key, info.checksum, info.file_size);
+
+        Ok(entry)
+    }
+
+    fn write_all(&self, key: FileKey, bytes: &[u8]) -> io::Result<FileEntry> {
+        if let Err(primary_error) = self.write_all_inner(key, bytes) {
+            return if let Err(secondary_error) = self.remove_file(key) {
+                Err(propagate_errors(primary_error, secondary_error))
+            } else {
+                Err(primary_error)
+            };
+        }
+
+        let checksum = self.compute_raw_checksum(bytes);
+        let entry = self.register_file_write(key, checksum, bytes.len() as u64);
+
+        Ok(entry)
+    }
+
+    fn remove_file(&self, key: FileKey) -> io::Result<()> {
+        let path = self.get_file_path(key);
+        self.register_file_removal(key);
+
+        match std::fs::remove_file(path) {
+            Err(e) if e.kind() == ErrorKind::NotFound => Ok(()),
+            other => other,
+        }
+    }
+
+    fn compute_file_info(&self, key: FileKey) -> io::Result<FileInfo> {
+        let file = self.get_open_file(key)?;
+        calculate_file_info(&file)
+    }
+}
+
+fn not_found_error() -> io::Error {
+    io::Error::new(ErrorKind::NotFound, "file not found")
+}
+
+fn propagate_errors(err1: io::Error, err2: io::Error) -> io::Error {
+    io::Error::new(
+        err1.kind(),
+        format!(
+            "Failed to complete IO operation: {}, during the handling of this error, another \
+             error occurred: {}",
+            err1, err2
+        ),
+    )
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct FileInfo {
+    pub checksum: u32,
+    pub file_size: u64,
+}
+
+fn calculate_file_info(file: &File) -> io::Result<FileInfo> {
+    let mut hasher = crc32fast::Hasher::new();
+
+    let file_size = file.metadata()?.len();
+    let mut cursor = 0;
+
+    let mut buffer = [0; 32 << 10];
+    while cursor < file_size {
+        let n = file.read_at(&mut buffer, cursor)?;
+
+        if n == 0 {
+            break;
+        }
+
+        hasher.write(&buffer[..n]);
+        cursor += n as u64;
+    }
+
+    Ok(FileInfo {
+        checksum: hasher.finalize(),
+        file_size,
+    })
+}
+
+fn read_to_fill_buff(file: &File, mut buf: &mut [u8], mut offset: u64) -> io::Result<()> {
```

Comment:
I don't like the name/the implem/the contract here, but I am not sure which whay it should be fixed.

Let me explain.

The implem is needlessly complex, and I have literally no clue how many times seek is called here.
I would be tempted to replace it to seek + read_exact...

However this function has a different behavior if the file finishes earler...
I suspect this is voluntary... Yet the name of the function does not help knowing this important detail and the prototype suggest otherwise.

For instance it does not return how much was read.
Maybe the caller saves the day by checking the integrity of the data in the buf, but this is not exactly pretty.